### PR TITLE
added new regex section to readme.md

### PR DESCRIPTION
**Pull request by tanberry**

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# LogDNA Agent
+git branch# LogDNA Agent
```
Comment (jorgebay) [inline docs/README.md:1]: Probably unintended :)

File: docs/README.md [hunk 2]
```diff
@@ -29,6 +29,7 @@ The LogDNA agent is a resource-efficient log collection client that forwards log
   * [Configuring Lookback](#configuring-lookback)
   * [Configuring Journald](#configuring-journald)
   * [Configuring Kubernetes Events](#configuring-events)
+  * [Configuring regex for redaction and exclusion or inclusion](#configuring-regex-for-redaction-and-exclusion-or-inclusion)
```
Comment (jorgebay) [inline docs/README.md:32]: The anchor needs updating :)

File: docs/README.md [hunk 3]
```diff
@@ -212,6 +213,22 @@ __Note:__ The default option is `never`.
...
+### Configuring regex for redaction and exclusion/inclusion
+
+You can define rules, using **regex** (regular expressions), to control what log data is ingested:
+ * include *only* specific log lines (with LOGDNA_LINE_INCLUSION_REGEX)
+ * exclude specific log lines (with LOGDNA_LINE_EXCLUSION_REGEX)
+ * redact parts of a log line (with LOGDNA_REDACT_REGEX)
+
+For example, you can identify certain types of logs that are noisy and not needed at all, and then write a regex pattern to match those files and preclude them from ingestion. Conversely, you can use regex expressions to match the log lines that you DO want in to include, and ingest only those log lines. Additionally, you can use the environment variable LOGDNA_REDACT_REGEX to remove certain parts of a log line. Any redacted data is replaced with [REDACTED].
+
+To access our library of common regex patterns, refer to [our regex library documentation](REGEX.md).
```
Comment (jorgebay) [inline docs/README.md:225]: Makes sense, I'll send a PR with patterns using the same file name.

File: docs/README.md [hunk 4]
```diff
@@ -212,6 +213,22 @@ __Note:__ The default option is `never`.
 
 > :warning: Due to a ["won't fix" bug in the Kubernetes API](https://github.com/kubernetes/kubernetes/issues/41743), the LogDNA agent collects events from the entire cluster, including multiple nodes. To prevent duplicate logs when running multiple pods, the LogDNA agent pods defer responsibilty of capturing events to the oldest pod in the cluster. If that pod is down, the next oldest LogDNA agent pod will take over responsibility and continue from where the previous pod left off.
 
+### Configuring regex for redaction and exclusion/inclusion
+
+You can define rules, using **regex** (regular expressions), to control what log data is ingested:
+ * include *only* specific log lines (with LOGDNA_LINE_INCLUSION_REGEX)
```
Comment (jorgebay) [inline docs/README.md:219]: Maybe we can use backticks for env var names, wdyt? like `LOGDNA_LINE_INCLUSION_REGEX`

File: docs/README.md [hunk 5]
```diff
@@ -212,6 +213,22 @@ __Note:__ The default option is `never`.
...
+ * exclude specific log lines (with LOGDNA_LINE_EXCLUSION_REGEX)
+ * redact parts of a log line (with LOGDNA_REDACT_REGEX)
+
+For example, you can identify certain types of logs that are noisy and not needed at all, and then write a regex pattern to match those files and preclude them from ingestion. Conversely, you can use regex expressions to match the log lines that you DO want in to include, and ingest only those log lines. Additionally, you can use the environment variable LOGDNA_REDACT_REGEX to remove certain parts of a log line. Any redacted data is replaced with [REDACTED].
+
+To access our library of common regex patterns, refer to [our regex library documentation](REGEX.md).
+
+Notes:
+   * Inclusion rules overwrite exclusion rules
+   * To preclude entire log files from being monitored, use the LOGDNA_EXCLUSION_RULES or the LOGDNA_EXCLUSION_REGEX_RULES environment variable.
```
Comment (jorgebay) [inline docs/README.md:229]: Good idea to remind the user about file exclusion.

(TIL _preclude_)

File: docs/README.md [hunk 6]
```diff
@@ -212,6 +213,22 @@ __Note:__ The default option is `never`.
...
+ * include *only* specific log lines (with LOGDNA_LINE_INCLUSION_REGEX)
+ * exclude specific log lines (with LOGDNA_LINE_EXCLUSION_REGEX)
+ * redact parts of a log line (with LOGDNA_REDACT_REGEX)
+
+For example, you can identify certain types of logs that are noisy and not needed at all, and then write a regex pattern to match those files and preclude them from ingestion. Conversely, you can use regex expressions to match the log lines that you DO want in to include, and ingest only those log lines. Additionally, you can use the environment variable LOGDNA_REDACT_REGEX to remove certain parts of a log line. Any redacted data is replaced with [REDACTED].
+
+To access our library of common regex patterns, refer to [our regex library documentation](REGEX.md).
+
+Notes:
+   * Inclusion rules overwrite exclusion rules
```
Comment (jorgebay) [inline docs/README.md:228]: This is not entirely correct.

For a line to be ingested, it should match all inclusion rules (if any) and not match any exclusion rule.

For example, if we use `"warning"` as inclusion regex and `"debug"` as exclusion regex (as in, "I want only warnings when those are not included in debug messages"), the following will occur:

- "warning about something" -> will match
- "this is a debug message about a warning" -> will not match
- "something" -> will not match (doesn't match inclusion)

If we want to summarize it,  we can say that exclusion rules overwrite inclusion rules.

File: docs/README.md [hunk 7]
```diff
@@ -212,6 +213,22 @@ __Note:__ The default option is `never`.
...
+   * To preclude entire log files from being monitored, use the LOGDNA_EXCLUSION_RULES or the LOGDNA_EXCLUSION_REGEX_RULES environment variable.
+   * Note that we use commas as separators for environment variable values, making it not possible to use the comma character (,) as a valid value. We are addressing this limitation in upcoming versions. If you need to use the comma character in a regular expression, use the unicode character reference: `\u002C`, for example: `hello\u002C world` matches `hello, world`.
```
Comment (jorgebay) [inline docs/README.md:230]: We could add a separate bullet point about case sensitivity, something like:

> All regular expressions are case sensitive by default. If you don't want to differentiate between upper and lower-case letters, use non-capturing groups with a flag: `(?flags:exp)`, for example: (?i:my_case_insensitive_regex)

**Review by jorgebay — COMMENTED** [review]
It's great to have a dedicated section for this.

Sorry about the number of remarks, it's a _meaty_ topic :)

File: docs/README.md (hunk 2 above)
Comment (tanberry) [reply docs/README.md:32]: I wasn't sure how to handle the special character (/) in the title...so I just changed the heading to match this link in ToC. But we can make heading whatever we want, let me know if we need to tweak.

File: docs/README.md (hunk 4 above)
Comment (tanberry) [reply docs/README.md:219]: yep, thanks, done.

File: docs/README.md (hunk 6 above)
Comment (tanberry) [reply docs/README.md:228]: Glad you caught that, thanks. Fixed it.

Comment (tanberry) [reply docs/README.md:228]: Oh... I wonder if I meant to say "Inclusion rules **override** (not overwrite) exclusion rules." EIther way, I reworded it as you suggested, that's better.

**Review by jorgebay — APPROVED** [review]
Looks great! 👍

**Review by c-nixon — APPROVED** [review]
This looks great to me!

**Pull request merged**